sql-plugin/src/main/scala/com/nvidia/spark/rapids/RmmRapidsRetryIterator.scala

```diff
@@ -0,0 +1,393 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.nvidia.spark.rapids
+
+import scala.collection.mutable
+
+import com.nvidia.spark.rapids.RapidsPluginImplicits._
+import com.nvidia.spark.rapids.jni.{RetryOOM, RmmSpark, SplitAndRetryOOM}
+
+object RmmRapidsRetryIterator extends Arm {
+
+  /**
+   * withRetry for Iterator[T]. This helper calls a function `fn` as it takes
+   * elements from the iterator given in `input`, and it can retry the work in `fn`,
+   * and optionally split items into smaller chunks. The `splitPolicy` function must
+   * close the item passed to it. The resulting iterator may or
+   * may not have the same number of elements as the source iterator.
+   *
+   * This function will close the elements of `input` as `fn` is successfully
+   * invoked. If the iterator `input` is not empty, in the event of an unhandled
+   * exception, it is closed entirely by `withRetry`.
+   *
+   * @param input an iterator of T
+   * @param splitPolicy a function that can split an item of type T into a Seq[T]. The split
+   *                    function must close the item passed to it.
+   * @param fn the work to perform. Takes T and produces an output K
+   * @tparam T element type that must be AutoCloseable
+   * @tparam K `fn` result type
+   * @return an iterator of K
+   */
+  def withRetry[T <: AutoCloseable, K](
```

Comment:
Can we add some explanation that even though T is AutoClosable, we actually expect T in most cases to be a SpillableColumnarBatch, and that whatever T is it should have the ability so that all of the data from T that is on the GPU is spillable at the moment of rollback before the retry.